kernel/src/schema.rs

```diff
@@ -647,61 +647,73 @@ impl Display for DataType {
 /// The provided `recurse_into_xxx` methods encapsulate the boilerplate work of recursing into the
 /// child schema elements of each schema element. Implementations can call these as needed but will
 /// generally not need to override them.
-pub trait SchemaTransform {
+pub trait SchemaTransform<'a> {
     /// Called for each primitive encountered during the schema traversal.
-    fn transform_primitive<'a>(
-        &mut self,
-        ptype: Cow<'a, PrimitiveType>,
-    ) -> Option<Cow<'a, PrimitiveType>> {
-        Some(ptype)
+    fn transform_primitive(&mut self, ptype: &'a PrimitiveType) -> Option<Cow<'a, PrimitiveType>> {
+        Some(Cow::Borrowed(ptype))
     }
 
     /// Called for each struct encountered during the schema traversal. Implementations can call
     /// [`Self::recurse_into_struct`] if they wish to recursively transform the struct's fields.
-    fn transform_struct<'a>(&mut self, stype: Cow<'a, StructType>) -> Option<Cow<'a, StructType>> {
+    fn transform_struct(&mut self, stype: &'a StructType) -> Option<Cow<'a, StructType>> {
         self.recurse_into_struct(stype)
     }
 
     /// Called for each struct field encountered during the schema traversal. Implementations can
     /// call [`Self::recurse_into_struct_field`] if they wish to recursively transform the field's
     /// data type.
-    fn transform_struct_field<'a>(
-        &mut self,
-        field: Cow<'a, StructField>,
-    ) -> Option<Cow<'a, StructField>> {
+    fn transform_struct_field(&mut self, field: &'a StructField) -> Option<Cow<'a, StructField>> {
         self.recurse_into_struct_field(field)
     }
 
     /// Called for each array encountered during the schema traversal. Implementations can call
     /// [`Self::recurse_into_array`] if they wish to recursively transform the array's element type.
-    fn transform_array<'a>(&mut self, atype: Cow<'a, ArrayType>) -> Option<Cow<'a, ArrayType>> {
+    fn transform_array(&mut self, atype: &'a ArrayType) -> Option<Cow<'a, ArrayType>> {
         self.recurse_into_array(atype)
     }
 
+    /// Called for each array element encountered during the schema traversal. Implementations can
+    /// call [`Self::transform`] if they wish to recursively transform the array element type.
+    fn transform_array_element(&mut self, etype: &'a DataType) -> Option<Cow<'a, DataType>> {
+        self.transform(etype)
```

Comment:
curious: why do we have these transform_array_element/map_key/map_val that just wrap `transform`?

Reply:
They're missing hook points this PR adds. The nested data skipping and column mapping PR will use them to improve its error messages (because column mapping transformation needs to recurse through maps and arrays, and if something goes wrong inside them, the error message should make it clear where the problem was.

Reply:
thanks :)